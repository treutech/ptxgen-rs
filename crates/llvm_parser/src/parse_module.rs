use anyhow::{Context, Result};
use llvm_ir::Module;
use std::fs;
use std::path::Path;

/// Parse an LLVM IR (.ll) file into an `llvm_ir::Module`
pub fn parse_module<P: AsRef<Path>>(path: P) -> Result<Module> {
    let ll_text = fs::read_to_string(&path)
        .with_context(|| format!("Failed to read file: {}", path.as_ref().display()))?;

    Module::from_ir_str(&ll_text)
        .map_err(|e| anyhow::anyhow!(e))
        .with_context(|| format!("Failed to parse LLVM IR in file: {}", path.as_ref().display()))
}
