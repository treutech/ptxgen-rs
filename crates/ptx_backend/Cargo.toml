
[package]
name = "ptx-backend"
version = "0.1.0"
edition = "2024"
license = "Apache-2.0"
authors = ["Raul Estrada <restrada@treutech.io>"]
homepage = "https://treutech.io/ptxgen-rs"
repository = "https://github.com/treutech/ptxgen-rs"

[dependencies]
ir-model = { path = "../ir_model" }
llvm-parser = { path = "../llvm_parser" }
clap = { version = "4.5.41", features = ["derive"] }
assert_cmd = "2.0.17"
anyhow = "1.0.98"
llvm-ir = "0.11.3"

[[bin]]
name = "ptx-backend"
path = "src/main.rs"

[dev-dependencies]
insta = "1.43.1"
