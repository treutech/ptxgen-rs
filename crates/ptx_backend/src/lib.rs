// Copyright 2025 Raul Estrada <restrada@treutech.io>
// SPDX-License-Identifier: Apache-2.0
//
// This file is part of the PTXGEN-RS project by Treu Technologies.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

mod utils;
use crate::utils::{clean_operand, get_register_type};
use ir_model::Instruction;
use std::collections::HashSet;

pub fn lower_function(name: &str, all_instrs: &[(String, Vec<Instruction>)]) -> Vec<String> {
    let mut output = vec![];

    output.push(format!("// Function: {}", name));
    output.push(emit_header());

    let flat_instrs: Vec<&Instruction> = all_instrs
        .iter()
        .flat_map(|(_, instrs)| instrs.iter())
        .collect();

    output.push(declare_registers(&flat_instrs));
    output.push(format!(".entry {} {{", name));

    for (block_name, instrs) in all_instrs {
        output.push(format!("    // Block: {}", block_name));
        for instr in instrs {
            let line = to_ptx(instr);
            output.push(format!("    {}", line));
        }
    }

    output.push("}".into());
    output
}

pub fn lower_to_ptx(instrs: &[&Instruction]) -> Vec<String> {
    let mut output = vec![];

    output.push("// Function: saxpy".into());
    output.push(emit_header());
    output.push(declare_registers(instrs));
    output.push(".entry saxpy {".into());

    for instr in instrs {
        output.push(to_ptx(instr));
    }

    output.push("}".into());
    output
}

fn emit_header() -> String {
    [
        ".version 7.0",
        ".target sm_75",
        ".address_size 64",
        "".into(),
    ]
    .join("\n")
}

pub fn declare_registers(instrs: &[&Instruction]) -> String {
    let mut f32_regs = HashSet::new();
    let mut s32_regs = HashSet::new();
    let mut pred_regs = HashSet::new();

    for instr in instrs {
        for name in instr.used_operands() {
            let reg = clean_operand(&name);
            match get_register_type(instr, &name) {
                Some("f32") => {
                    f32_regs.insert(reg.to_string());
                }
                Some("s32") => {
                    s32_regs.insert(reg.to_string());
                }
                Some("pred") => {
                    pred_regs.insert(reg.to_string());
                }
                _ => {}
            }
        }
    }

    let mut out = vec![];
    if !f32_regs.is_empty() {
        out.push(format!(
            ".reg .f32 {};",
            f32_regs
                .iter()
                .map(|r: &String| format!("%{}", r))
                .collect::<Vec<String>>()
                .join(", ")
        ));
    }
    if !s32_regs.is_empty() {
        out.push(format!(
            ".reg .s32 {};",
            s32_regs
                .iter()
                .map(|r: &String| format!("%{}", r))
                .collect::<Vec<String>>()
                .join(", ")
        ));
    }
    if !pred_regs.is_empty() {
        out.push(format!(
            ".reg .pred {};",
            pred_regs
                .iter()
                .map(|r: &String| format!("%{}", r))
                .collect::<Vec<String>>()
                .join(", ")
        ));
    }
    out.push(String::new()); // for newline
    out.join("\n")
}

pub fn to_ptx(instr: &Instruction) -> String {
    match instr {
        Instruction::FMul { dst, lhs, rhs } => {
            format!(
                "    fmul.f32 {}, {}, {};",
                clean_operand(dst),
                clean_operand(lhs),
                clean_operand(rhs)
            )
        }
        Instruction::FAdd { dst, lhs, rhs } => {
            format!(
                "    fadd.f32 {}, {}, {};",
                clean_operand(dst),
                clean_operand(lhs),
                clean_operand(rhs)
            )
        }
        Instruction::Load { dst, src } => {
            format!(
                "    ld.global.f32 {}, {};",
                clean_operand(dst),
                clean_operand(src)
            )
        }
        Instruction::Store { dst, value } => {
            format!(
                "    st.global.f32 {}, {};",
                clean_operand(dst),
                clean_operand(value)
            )
        }
        Instruction::Add { dst, lhs, rhs } => {
            format!(
                "    add.s32 {}, {}, {};",
                clean_operand(dst),
                clean_operand(lhs),
                clean_operand(rhs)
            )
        }
        Instruction::ICmp { dst, lhs, rhs, .. } => {
            format!(
                "    setp.lt.s32 {}, {}, {};",
                clean_operand(dst),
                clean_operand(lhs),
                clean_operand(rhs)
            )
        }
        Instruction::Alloca { dst, .. } => {
            format!("    // local stack allocation: {}", clean_operand(dst))
        }
        Instruction::GetElementPtr { dst, base, index } => {
            format!(
                "    // address calc: {} = {}[{}]",
                clean_operand(dst),
                clean_operand(base),
                clean_operand(index)
            )
        }
        Instruction::Unhandled(s) => format!("    // unhandled: {}", s),
        _ => String::from("    // unsupported instruction"),
    }
}
