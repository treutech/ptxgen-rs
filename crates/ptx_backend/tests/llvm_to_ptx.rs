use llvm_parser::parse_module::parse_module;
use ptx_backend::lower_to_ptx;

#[test]
fn test_saxpy_ptx_output() {

    let path = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("examples")
        .join("saxpy.ll");
    println!("Reading from path: {}", path.display());

    let module = parse_module(&path).expect("Failed to parse module");

    let mut actual = String::new();

    for func in module.functions {
        actual.push_str(&format!("// Function: {}\n", func.name));
        for block in func.basic_blocks {
            actual.push_str(&format!("// Block: {}\n", block.name));
            let instrs = block
                .instrs
                .iter()
                .map(llvm_parser::convert::lower)
                .collect::<Vec<_>>();
            for line in lower_to_ptx(&instrs) {
                actual.push_str(&format!("{}\n", line));
            }
        }
    }

    insta::assert_snapshot!("saxpy_ptx", actual);
}
