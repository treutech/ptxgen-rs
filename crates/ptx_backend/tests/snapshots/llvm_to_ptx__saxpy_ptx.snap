---
source: crates/ptx_backend/tests/llvm_to_ptx.rs
expression: actual
---
.version 7.0
.target sm_75
.address_size 64

// Function: saxpy
// Block: %entry
// Block: %loop
// Block: %body
// Block: %exit
.reg .f32 %y_, %a, %out_, %xval, %yval, %next, %ax, %0, %i, %idx, %res, %x_;
.reg .s32 %next, %1, %idx, %n;
.reg .pred %cmp;
.entry saxpy {
        // local stack allocation: i
        st.global.f32 i, 0;
        ld.global.f32 idx, i;
        setp.lt.s32 cmp, idx, n;
        // address calc: x_ = x[idx]
        // address calc: y_ = y[idx]
        // address calc: out_ = out[idx]
        ld.global.f32 xval, x_;
        ld.global.f32 yval, y_;
        fmul.f32 ax, a, xval;
        fadd.f32 res, ax, yval;
        st.global.f32 out_, res;
        add.s32 next, idx, 1;
        st.global.f32 i, next;
}
